### PR TITLE
first commit

### PR DESCRIPTION
**Pull request by RuslanIbIsh**

### CODE REVIEW

**Review by Ilya-brinchuk — APPROVED** [review]

File: src/main/java/com/iri/movietickets/controller/CinemaHallController.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("cinemahalls")
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/CinemaHallController.java:16]: please see mapping in the task description

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.stream.Collectors;
+import org.springframework.format.annotation.DateTimeFormat;
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestParam;
+
+public class MovieSessionController {
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/MovieSessionController.java:18]: does this controller work?

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                .collect(Collectors.toList());
+    }
+
+    @PostMapping
+    public void create(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        movieSessionService.add(movieSessionMapper.convertFromDto(movieSessionRequestDto));
+    }
+
+    @PutMapping
+    public void update(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/MovieSessionController.java:44]: let's pass id as path variable here as well

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+        movieSessionService.update(movieSessionMapper.convertFromDto(movieSessionRequestDto));
+    }
+
+    @DeleteMapping
+    public void delete(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/MovieSessionController.java:49]: let's pass only id as path variable here

File: src/main/java/com/iri/movietickets/dao/MovieSessionDao.java [hunk 5]
```diff
@@ -8,4 +8,8 @@
     List<MovieSession> findAvailableSessions(Long movieId, LocalDate date);
 
     MovieSession add(MovieSession session);
+
+    MovieSession update(MovieSession movieSession);
+
+    MovieSession delete(MovieSession movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/dao/MovieSessionDao.java:14]:
```suggestion
    void delete(Long movieSessionId);
```

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java [hunk 6]
```diff
@@ -62,4 +62,51 @@ public MovieSession add(MovieSession movieSession) {
...
+    }
+
+    @Override
+    public MovieSession delete(MovieSession movieSession) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.delete(movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java:97]: let's use load() method to get the session by id before deleting it
or write our own HQL query

File: src/main/java/com/iri/movietickets/model/dto/CinemaHallResponseDto.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package com.iri.movietickets.model.dto;
+
+public class CinemaHallResponseDto {
+    private Long id;
+    private String description;
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/model/dto/CinemaHallResponseDto.java:5]: what about capacity?

File: src/main/java/com/iri/movietickets/model/dto/MovieResponseDto.java [hunk 8]
```diff
@@ -0,0 +1,22 @@
+package com.iri.movietickets.model.dto;
+
+public class MovieResponseDto {
+    private Long id;
+    private String title;
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/model/dto/MovieResponseDto.java:5]: what about description?

File: src/main/java/com/iri/movietickets/model/dto/MovieSessionRequestDto.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
+package com.iri.movietickets.model.dto;
+
+public class MovieSessionRequestDto {
+    private Long movieSessionId;
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/model/dto/MovieSessionRequestDto.java:4]: let's not pass this id in requestDto - instead pass it as path variable in `update()` method

File: src/main/java/com/iri/movietickets/service/impl/MovieServiceImpl.java [hunk 10]
```diff
@@ -23,4 +24,11 @@ public Movie add(Movie movie) {
     public List<Movie> getAll() {
         return movieDao.getAll();
     }
+
+    @Override
+    public Movie get(Long id) {
+        return movieDao.get(id).orElseThrow(() ->
+                new DataProcessingException("Movie session service could not get"
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/service/impl/MovieServiceImpl.java:31]:
```suggestion
                new DataProcessingException("Movie service could not get"
```

File: src/main/java/com/iri/movietickets/service/mapper/MovieSessionMapper.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        movieSessionResponseDto.setMovieSessionId(movieSession.getId());
+        movieSessionResponseDto.setMovieTitle(movieSession.getMovie().getTitle());
+        movieSessionResponseDto.setCinemaHallId(movieSession.getCinemaHall().getId());
+        movieSessionResponseDto.setShowTime(movieSession.getShowTime().toString());
+        return movieSessionResponseDto;
+    }
+
+    public MovieSession convertFromDto(MovieSessionRequestDto movieSessionRequestDto) {
+        MovieSession movieSession = new MovieSession();
+        movieSession.setId(movieSessionRequestDto.getMovieSessionId());
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/service/mapper/MovieSessionMapper.java:34]: let's remove this line from here

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/iri/movietickets/controller/CinemaHallController.java (hunk 1 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/CinemaHallController.java:16]: done

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java (hunk 2 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/MovieSessionController.java:18]: done

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java (hunk 3 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/MovieSessionController.java:44]: done

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java (hunk 4 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/MovieSessionController.java:49]: done

File: src/main/java/com/iri/movietickets/dao/MovieSessionDao.java (hunk 5 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/dao/MovieSessionDao.java:14]: done

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java (hunk 6 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java:97]: done

File: src/main/java/com/iri/movietickets/model/dto/CinemaHallResponseDto.java (hunk 7 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/model/dto/CinemaHallResponseDto.java:5]: done

File: src/main/java/com/iri/movietickets/model/dto/MovieResponseDto.java (hunk 8 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/model/dto/MovieResponseDto.java:5]: done

File: src/main/java/com/iri/movietickets/model/dto/MovieSessionRequestDto.java (hunk 9 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/model/dto/MovieSessionRequestDto.java:4]: done

File: src/main/java/com/iri/movietickets/service/impl/MovieServiceImpl.java (hunk 10 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/service/impl/MovieServiceImpl.java:31]: done

File: src/main/java/com/iri/movietickets/service/mapper/MovieSessionMapper.java (hunk 11 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/service/mapper/MovieSessionMapper.java:34]: done

File: src/main/java/com/iri/movietickets/controller/CinemaHallController.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+import java.util.List;
+import java.util.stream.Collectors;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("cinema-halls")
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/CinemaHallController.java:16]:
```suggestion
@RequestMapping("/cinema-halls")
```

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+
+@Controller
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/MovieSessionController.java:21]: let's check if `getAvailable()` method returns the result as json

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+    @PostMapping
+    public void create(@RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        movieSessionService.add(movieSessionMapper.convertFromDto(movieSessionRequestDto));
+    }
+
+    @PutMapping("/{id}")
+    public void update(@PathVariable Long id,
+                       @RequestBody MovieSessionRequestDto movieSessionRequestDto) {
+        movieSessionService.update(id, movieSessionMapper.convertFromDto(movieSessionRequestDto));
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/controller/MovieSessionController.java:51]:
```suggestion
        // convert dto to entity
        // set id
        movieSessionService.update(movieSession);
```

File: src/main/java/com/iri/movietickets/dao/MovieSessionDao.java [hunk 15]
```diff
@@ -8,4 +8,8 @@
     List<MovieSession> findAvailableSessions(Long movieId, LocalDate date);
 
     MovieSession add(MovieSession session);
+
+    MovieSession update(Long id, MovieSession movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/dao/MovieSessionDao.java:12]:
```suggestion
    MovieSession update(MovieSession movieSession);
```

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java [hunk 16]
```diff
@@ -62,4 +62,60 @@ public MovieSession add(MovieSession movieSession) {
...
+    public MovieSession update(Long id, MovieSession movieSession) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            Query updateQuery = session.createQuery("update MovieSession ms "
+                    + "set ms.showTime = :showTime, "
+                    + "ms.movie = :movie, ms.cinemaHall = :cinemaHall "
+                    + "where ms.id = :id", MovieSession.class);
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java:76]: we can just use `session.update()` here

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java [hunk 17]
```diff
@@ -62,4 +62,60 @@ public MovieSession add(MovieSession movieSession) {
...
+    }
+
+    @Override
+    public void delete(Long id) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            Query deleteQuery = session.createQuery("delete MovieSession m "
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java:105]: is the syntax ok?

File: src/main/java/com/iri/movietickets/service/MovieSessionService.java [hunk 18]
```diff
@@ -8,4 +8,8 @@
     List<MovieSession> findAvailableSessions(Long movieId, LocalDate date);
 
     MovieSession add(MovieSession session);
+
+    MovieSession update(Long id, MovieSession movieSession);
```
Comment (KseniiaMakarova) [inline src/main/java/com/iri/movietickets/service/MovieSessionService.java:12]:
```suggestion
    MovieSession update(MovieSession movieSession);
```

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/iri/movietickets/controller/CinemaHallController.java (hunk 12 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/CinemaHallController.java:16]: done

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java (hunk 13 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/MovieSessionController.java:21]: done

File: src/main/java/com/iri/movietickets/controller/MovieSessionController.java (hunk 14 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/controller/MovieSessionController.java]: done

File: src/main/java/com/iri/movietickets/dao/MovieSessionDao.java (hunk 15 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/dao/MovieSessionDao.java]: done

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java (hunk 17 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java:97]: done

File: src/main/java/com/iri/movietickets/service/MovieSessionService.java (hunk 18 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/service/MovieSessionService.java]: done

File: src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java (hunk 16 above)
Comment (RuslanIbIsh) [reply src/main/java/com/iri/movietickets/dao/impl/MovieSessionDaoImpl.java]: done

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**